### PR TITLE
[T4PUB-472] Re-construct the structure of ota_client.py

### PR DESCRIPTION
**Pull request by Bodong-Yang**
Previous implementation of OtaClient is a monolithic huge class containing all methods and attributes related to ota_client, which make the code base hard to read and maintain.

This PR re-design the structure of the OtaClient, introducing modularization structure to the implementation.

### CODE REVIEW

File: app/ota_client.py [hunk 1]
```diff
@@ -1292,8 +1320,158 @@ def _rollback(self):
...
+    _mount_point: str = None
+    _inform_update_error: Callable = None
+
+    def reboot(self):
+        """
+        main entry for OtaReboot
+        """
+        if self.get_ota_status() == OtaStatusString.PREPARED_STATE:
+            # switch reboot
+            if not self._grub_ctl.prepare_grub_switching_reboot(
```
Comment (obi-t4) [inline app/ota_client.py:1349]:
```suggestion
            if not self.prepare_grub_switching_reboot(
```

File: app/ota_client.py [hunk 2]
```diff
@@ -1292,8 +1320,158 @@ def _rollback(self):
...
+                # inform error
+                self._inform_update_error("Switching bank failed!")
+                # set 'NORMAL' state
+                self._ota_status.set_ota_status(OtaStatusString.NORMAL_STATE)
+                _unmount_bank(self._mount_point)
+                return False
+            #
+            # -----------------------------------------------------------
+            # set 'SWITCHA/SWITCHB' state
+            next_bank = self._grub_ctl.get_next_bank()
```
Comment (obi-t4) [inline app/ota_client.py:1361]:
```suggestion
            next_bank = self.get_next_bank()
```

File: app/ota_client.py [hunk 3]
```diff
@@ -1292,8 +1320,158 @@ def _rollback(self):
...
+            next_state = self._get_switch_status_for_reboot(next_bank)
+            self._ota_status.set_ota_status(next_state)
+        #
+        # reboot
+        #
+        self._grub_ctl.reboot()
```
Comment (obi-t4) [inline app/ota_client.py:1367]:
```suggestion
        self.reboot()
```

File: app/ota_client.py [hunk 4]
```diff
@@ -1167,65 +1203,55 @@ def update(self, ecu_update_info):
         # set 'METADATA' state
```
Comment (obi-t4) [inline app/ota_client.py:1203]: `self._ota_status.set_ota_status(OtaStatusString.NORMAL_STATE)`
should be `self.set_ota_status(OtaStatusString.NORMAL_STATE)`

File: app/ota_client.py [hunk 5]
```diff
@@ -1167,65 +1203,55 @@ def update(self, ecu_update_info):
...
             # set 'NORMAL' state
-            self._ota_status.set_ota_status(OtaStatusString.NORMAL_STATE)
+            self.set_ota_status(OtaStatusString.NORMAL_STATE)
             _unmount_bank(self._mount_point)
             return False
         #
         # -----------------------------------------------------------
         # set 'PREPARED' state
+        logger.debug(f"finished update, set ota_status")
         self._ota_status.set_ota_status(OtaStatusString.PREPARED_STATE)
```
Comment (obi-t4) [inline app/ota_client.py:1220]:
```suggestion
        self.set_ota_status(OtaStatusString.PREPARED_STATE)
```

File: app/ota_client.py [hunk 6]
```diff
@@ -1007,155 +1020,178 @@ def _setup_regular_files(self, target_dir: Path):
...
+    _mount_point: Path = None
 
     def update(self, ecu_update_info):
         """
         OTA update execution
         """
         # -----------------------------------------------------------
         # set 'UPDATE' state
+        logger.debug(f"init update")
         self._ota_status.set_ota_status(OtaStatusString.UPDATE_STATE)
```
Comment (obi-t4) [inline app/ota_client.py:1181]:
```suggestion
        self.set_ota_status(OtaStatusString.UPDATE_STATE)
```

File: app/ota_client.py [hunk 7]
```diff
@@ -1292,8 +1320,158 @@ def _rollback(self):
             #
```
Comment (obi-t4) [inline app/ota_client.py:1320]: `self._ota_status.set_ota_status(next_state)`
should be `self.set_ota_status(next_state)`

File: app/ota_client.py [hunk 8]
```diff
@@ -1292,8 +1320,158 @@ def _rollback(self):
...
+        """
+        if self.get_ota_status() == OtaStatusString.PREPARED_STATE:
+            # switch reboot
+            if not self._grub_ctl.prepare_grub_switching_reboot(
+                self._boot_vmlinuz, self._boot_initrd
+            ):
+                # inform error
+                self._inform_update_error("Switching bank failed!")
+                # set 'NORMAL' state
+                self._ota_status.set_ota_status(OtaStatusString.NORMAL_STATE)
```
Comment (obi-t4) [inline app/ota_client.py:1355]:
```suggestion
                self.set_ota_status(OtaStatusString.NORMAL_STATE)
```

**Review by obi-t4 — COMMENTED** [review]
Please check comments :bow:

**Comment by Bodong-Yang** [comment]
Closed as we have new implementation.

**Pull request closed without merging**